keps/sig-storage/20190122-csi-inline-volumes.md

```diff
@@ -110,51 +83,16 @@ spec:
-
-* Persistent mode - the volume follows a persistent lifecycle independent of the pod where it is used.
-* Ephemeral mode - the volume follows the lifecycle of the pod where it is used.
-
-A driver may be able to support either PV/PVC-originated or pod spec originated volumes (inline).  The following is important to keep in mind:
-* When a volume originates from PV/PVC, the driver will receive all volume related CSI calls (normal operation).
-* When a volume originates from a pod spec (inline), the CSI driver will only receive a limited number of calls (see below).
-
-### Inline CSI lifecycle setting
-A CSI driver will have to indicate the inline volume lifecycle, that it supports, as either `persistent` or `ephemeral`.  This setting will be stored in a [`CSIDriver` configuration CRD](https://github.com/kubernetes/enhancements/issues/594) and will be checked at runtime by Kubelet and external CSI components for enforcement.  The setting is to inform the Kubelet and other CSI components how to process inline volume information for proper setup.
-
-#### Persistent inline mode
-For CSI drivers that indicate they support persistent inline volumes, the `CSIDriver.lifecycle` value should be set to `persistent` (default of if not provided).  When a driver is used to handle volumes from an inline context: 
-* Volumes are assumed to be originated from an inline pod spec.
-* The `volumeHandle` value must be generated ahead of time and manually entered in the pod spec.
-* The driver will only receive CSI calls related to storage opearations for *attach*, *mount*, *unmount*, *detach*.
-* The driver will not receive CSI calls for *porovision* and *deletion* of volumes.
-
-#### Ephemeral inline mode
-Ephemeral inline volumes are more restricted than their persistent counterparts.  The `CSIDriver.lifecycle = ephemeral` must be set to indicate support for inline ephemeral volumes which will follow the lifecycle of its pod.  For drivers indicating support for ephemral mode: 
-* Volumes are assumed to be originated from an inline pod spec.
-* The `volumeHandle` is not required and is ignored if provided in the pod spec.
-* CSI will internally generate a `volumeHandle` value which is passed to the driver.
-* Ephemeral volumes will only make CSI calls for *mount* and *unmount* volume operations.
-
-#### Omitting inline mode
-Omitting `CSIDriver.lifecycle` is an indication:
-* The driver can support PV/PVC originated volume (normal operation).
-* If the volume originates from a pod spec, it is assumed that its lifecycle is `persistent`.
+## Ephemeral inline volume proposal
+A CSI driver may be able to support either PV/PVC-originated or pod spec originated volumes. When a volume definition is embedded inside a pod spec, it is considered to be an `ephemeral inline` volume request and can only participate in *mount/unmount* volume operation calls.  Ephemeral inline volume requests have the following characteristics: 
```

Comment:
Previously we added a flag in CSIDriver CRD to ensure that the driver was capable of operating properly as an ephemeral driver. I think this is still important.

Reply:
What would the flag be used to gate?

Reply:
it would prevent calling mount/unmount on a CSI driver for an inline ephemeral volume source, if the driver did not support ephemeral volumes

otherwise we're assuming the driver will fail in a safe/performant/reliable way when given a bogus (generated) volumeHandle, which doesn't seem great.

Reply:
Ok to summarize, it would be used to gate the auto-generation of a volume handle during mount/unmount.  And this is to protect persistent drivers, not to protect ephemeral drivers.

Reply:
> And this is to protect persistent drivers, not to protect ephemeral drivers.

protection would go both ways, I'd imagine (ephemeral drivers wouldn't be eligible for being called to satisfy storageclass or PV-related things)

Reply:
hm ok, we would have to check this in every CSI driver callout, but it's doable.

Reply:
I think that we don't need the CRD.Lifecycle any longer. If CSIVolumeSource shows up in volume.Spec, it should be assumed that it is an ephemeral. Also, we have the `PodSecurityPolicy` to indicate which inline drivers are allowed. Now that only Ephemeral is supported we should take the opportunity to reduce configuration surface.

Reply:
Actually, I thought about this further.  `CSIDriver.Lifecycle` flag doesn't make sense anymore since inline will always be ephemeral.  However we probably still need something to indicate whether the driver support/operating as ephemeral driver.  While PSP will gate which ephemeral driver can run, it does not offer the sort of protection discussed in this thread above.

Reply:
I will update the KEP to introduce `CSIDriver.Spec.InlineSupported = {true|false}` which indicates the driver supports Inline volumes .  If `CSIDriver.InlineSupported = true` then `CSIDriver.Spec.AttachRequired` is assumed false causing attach phase to be skipped.

Reply:
I kind of like the PodInfoOnMountVersion *string field in CSIDriver:

It allows different versions of the protocol to be specified as supported. So, maybe we have it:
EphemeralVersion string = "v1" and then in the future, if we ever came up with some more advanced types of ephemeral (other types then just mount/umount) then we have a place to hang those off of?

Reply:
I don't know about "v1". It's not very clear what that means.  We could go back to the original idea of "ephemeral" and "persistent" strings, and we'll have to think of a new name of something else changes.

Drivers should still report their capabilities for provision, attach, resize, etc.  I'm not sure if we will need a new type string if we want to expand the scope of ephemeral volumes in the future.

Reply:
thats true.

Reply:
@msau42 OK. The flag will be a string where the current supported value is "ephemeral".  Future implementation may define a new value.  A value of "persistent" does not make sense right now since the code will not support that feature.

I will update the KEP again to reflect that and the code.

Reply:
We still do want persistent right?  Persistent supports provision, attach, resize, mount, etc, but does not support volume handle generation.  Ephemeral only supports mount/unmount and volume handle generation.

Reply:
@msau42 
It sounded like here (https://github.com/kubernetes/kubernetes/pull/74086#issuecomment-465813975) that persistent inline CSI would not be supported at all.  Maybe I am missing something.  Let me know what you think.  If you think we will come back (beta) to support persistent inline CSI, then I will word the KEP.

Reply:
Persistent inline is not supported. However, it may still be useful to distinguish between persistent and ephemeral drivers in general due to their different lifecycles

Reply:
@msau42 I see.
OK. I see your point. I will rename the flag to reflect that distinction `CSIDriver.Mode = {ephemeral|persistent}`.  Thanks for clarifying.